### PR TITLE
[inductor] Disable fp contraction and add option to use precise division

### LINKED ISSUE #101039: [torch.compile] the sum of `softmax` isn't `1` on cuda
### 🐛 Describe the bug

`torch.compile` returns the sum of `softmax` isn't equal to `1` on cuda

```py
import torch

torch.manual_seed(420)

class Model(torch.nn.Module):

    def __init__(self):
        super().__init__()
        self.softmax = torch.nn.Softmax(dim=-1)

    def forward(self, query, key):
        y = torch.matmul(query, key.transpose(-2, -1))
        z = y.div(1e-06)
        w = self.softmax(z)
        return w

device = 'cuda'
func = Model().to(device)

query = torch.randn(1, 10, 40).to(device)

key = torch.randn(1, 2, 40).to(device)

jit_func = torch.compile(func)

res1 = func(query, key) # without jit
print(res1)
# tensor([[[0., 1.],
#          [0., 1.],
#          [0., 1.],
#          [0., 1.],
#          [0., 1.],
#          [1., 0.],
#          [1., 0.],
#          [0., 1.],
#          [0., 1.],
#          [0., 1.]]], device='cuda:0')

res2 = jit_func(query, key)
print(res2)
# tensor([[[0.0000, 0.9804],
#          [0.0000, 1.0561],
#          [0.0000, 1.1190],
#          [0.0000, 0.9721],
#          [0.0000, 1.1999],
#          [0.9972, 0.0000],
#          [0.8769, 0.0000],
#          [0.0000, 1.0315],
#          [0.0000, 1.0398],
#          [0.0000, 0.8039]]], device='cuda:0')
```

### Versions

<details>
<summary>Click to expand</summary>

```
Collecting environment information...
PyTorch version: 2.1.0.dev20230503+cu118
Is debug build: False
CUDA used to build PyTorch: 11.8
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.1 LTS (x86_64)
GCC version: (Ubuntu 11.3.0-1ubuntu1~22.04) 11.3.0
Clang version: 14.0.0-1ubuntu1
CMake version: Could not collect
Libc version: glibc-2.35

Python version: 3.9.16 (main, Mar  8 2023, 14:00:05)  [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-5.19.5-051905-generic-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 11.5.119
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 3060
Nvidia driver version: 510.108.03
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Address sizes:                   46 bits physical, 48 bits virtual
Byte Order:                      Little Endian
CPU(s):                          24
On-line CPU(s) list:             0-23
Vendor ID:                       GenuineIntel
Model name:                      12th Gen Intel(R) Core(TM) i9-12900K
CPU family:                      6
Model:                           151
Thread(s) per core:              2
Core(s) per socket:              16
Socket(s):                       1
Stepping:                        2
CPU max MHz:                     6700.0000
CPU min MHz:                     800.0000
BogoMIPS:                        6374.40
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf tsc_known_freq pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault cat_l2 invpcid_single cdp_l2 ssbd ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid rdt_a rdseed adx smap clflushopt clwb intel_pt sha_ni xsaveopt xsavec xgetbv1 xsaves split_lock_detect avx_vnni dtherm ida arat pln pts hwp hwp_notify hwp_act_window hwp_epp hwp_pkg_req hfi umip pku ospke waitpkg gfni vaes vpclmulqdq tme rdpid movdiri movdir64b fsrm md_clear serialize pconfig arch_lbr ibt flush_l1d arch_capabilities
Virtualization:                  VT-x
L1d cache:                       640 KiB (16 instances)
L1i cache:                       768 KiB (16 instances)
L2 cache:                        14 MiB (10 instances)
L3 cache:                        30 MiB (1 instance)
NUMA node(s):                    1
NUMA node0 CPU(s):               0-23
Vulnerability Itlb multihit:     Not affected
Vulnerability L1tf:              Not affected
Vulnerability Mds:               Not affected
Vulnerability Meltdown:          Not affected
Vulnerability Mmio stale data:   Not affected
Vulnerability Retbleed:          Not affected
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:        Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:        Mitigation; Enhanced IBRS, IBPB conditional, RSB filling, PBRSB-eIBRS SW sequence
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Not affected

Versions of relevant libraries:
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.24.2
[pip3] pytorch-triton==2.1.0+7d1a95b046
[pip3] torch==2.1.0.dev20230503+cu118
[pip3] torchaudio==2.1.0.dev20230503+cu118
[pip3] torchvision==0.16.0.dev20230503+cu118
[conda] numpy                     1.24.2                   pypi_0    pypi
[conda] pytorch-triton            2.1.0+7d1a95b046          pypi_0    pypi
[conda] torch                     2.1.0.dev20230503+cu118          pypi_0    pypi
[conda] torchaudio                2.1.0.dev20230503+cu118          pypi_0    pypi
[conda] torchvision               0.16.0.dev20230503+cu118          pypi_0    pypi
```

</details>

cc @ezyang @msaroufim @wconstab @ngimel @bdhirsh @anijain2305 @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @soumith

### LINKED ISSUE #122260: fma can cause drastically worse precision in torch.compile/Triton
### 🐛 Describe the bug

```python
import torch
torch.set_default_device('cuda')

scale = torch.tensor(0.180336877703666687)
x = torch.tensor(1134139801600.000000)

def f(x, scale):
    max_scaled = x * scale
    return torch.exp(max_scaled - x * scale)

print(f(x, scale))
print(torch.compile(f)(x, scale))
>>> tensor(1., device='cuda:0')
>>> tensor(inf, device='cuda:0')
```

The root cause here is the same as here: https://github.com/pytorch/pytorch/issues/121558#issuecomment-2008359796

I think the general structure of this can show up in many cases, but in this case, this is a Triton issue. Here, Inductor actually CSE's the `x * scale` call.

The generated Triton code is
```python
@triton_heuristics.pointwise(
    size_hints=[1],
    filename=__file__,
    triton_meta={'signature': {0: '*fp32', 1: '*fp32', 2: '*fp32', 3: 'i32'}, 'device': 0, 'device_type': 'cuda', 'constants': {3: 1}, 'configs': [AttrsDescriptor(divisible_by_16=(0, 1, 2), equal_to_1=(3,), ids_of_folded_args=(3,), divisible_by_8=())]},
    inductor_meta={'autotune_hints': set(), 'kernel_name': 'triton_poi_fused_exp_mul_sub_0', 'mutated_arg_names': [], 'no_x_dim': False, 'backend_hash': 'cddd2cc4107921f6715d58b181fdb08b23055085471461101752ba6efb772ae1'},
    min_elem_per_thread=0
)
@triton.jit
def triton_(in_ptr0, in_ptr1, out_ptr0, xnumel, XBLOCK : tl.constexpr):
    xnumel = 1
    xoffset = tl.program_id(0) * XBLOCK
    xindex = xoffset + tl.arange(0, XBLOCK)[:]
    xmask = xindex < xnumel
    tmp0 = tl.load(in_ptr0 + (0))
    tmp1 = tl.broadcast_to(tmp0, [XBLOCK])
    tmp2 = tl.load(in_ptr1 + (0))
    tmp3 = tl.broadcast_to(tmp2, [XBLOCK])
    tmp4 = tmp1 * tmp3
    tmp5 = tmp4 - tmp4
    tmp6 = tl_math.exp(tmp5)
    tl.store(out_ptr0 + (tl.full([XBLOCK], 0, tl.int32)), tmp6, None)
```

Here, you would imagine that `tmp4 - tmp4` is 0. Unfortunately, Triton generates this PTX. 
```
        mul.f32         %f5, %f3, %f4;
        .loc    1 30 18
        neg.f32         %f6, %f5;
        fma.rn.f32      %f7, %f3, %f4, %f6;
```
Which duplicates the FMA again!

### Versions

N/A

cc @ezyang @msaroufim @bdhirsh @anijain2305 @zou3519 @chauhang @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire

### PR DESCRIPTION
**Pull request by peterbell10**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #122915
* __->__ #115435
* #123269

Fixes #122260

In the issue we have something like:
```python
(x / 1e-6) + (y / 1e-6)
```
which triton turns into
```python
fma(x, 1e6, y * 1e6)
```

Where we have one division calculated to float32 precision,
and the other calculated to infinite precision as part of the `fma`.
In the issue this leads to a bad normalization in softmax.

Here we disallow the "fp fusion" optimization which generates these fma instructions.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

### CODE REVIEW

**Comment by bertmaher** [comment]
Just curious, if we didn't use `fma` here, and instead computed `x*1e6 + y*1e6`, would we still get the wrong result here?  I'm kind of wondering if fma in general is a problem (sort of independently of non-precise division).  I believe the CPU backend for inductor actually disables fma fusion.

**Comment by peterbell10** [comment]
Yeah I'll give that a shot as well.

**Comment by desertfire** [comment]
I think it will be great if we can provide a torch.compile mode that turns off all the optimizations known to cause large numerical differences. When users find torch.compile produces an unacceptable numerical difference from eager, the first thing they could try is to switch to that special mode.

**Comment by peterbell10** [comment]
> I think it will be great if we can provide a torch.compile mode that turns off all the optimizations known to cause large numerical differences. When users find torch.compile produces an unacceptable numerical difference from eager, the first thing they could try is to switch to that special mode.

I suspect the biggest factor here is that fusions often prevent floating point values from being truncated to half precision. In one sense this is good as it leads to higher precision, but as this issue shows the mismatch in precision between two different paths to compute the same value can lead to numerical issues.

**Comment by peterbell10** [comment]
No significant swings on the benchmark results for disabling fp-contraction. Will have to dig into these torchbench accuracy failures though.

File: torch/_inductor/triton_heuristics.py [hunk 1]
```diff
@@ -339,6 +339,7 @@ def _precompile_config(self, cfg: Config, warm_cache_only_with_cc: Optional[int]
                 "num_warps": compile_meta["num_warps"],
                 "num_stages": compile_meta["num_stages"],
                 "debug": compile_meta["debug"],
+                "enable_fp_fusion": config.triton.enable_fp_fusion,
```
Comment (oulgen) [inline torch/_inductor/triton_heuristics.py:342]: @Chillee pointed out this me. I wonder how this will play with user defined triton kernels? Should we exclude those from this setting?

File: torch/_inductor/triton_heuristics.py [hunk 2]
```diff
@@ -339,6 +339,11 @@ def _precompile_config(self, cfg: Config, warm_cache_only_with_cc: Optional[int]
                 "num_warps": compile_meta["num_warps"],
                 "num_stages": compile_meta["num_stages"],
                 "debug": compile_meta["debug"],
+                "enable_fp_fusion": (
+                    config.triton.enable_fp_fusion
+                    or self.heuristic_type
+                    in (HeuristicType.TEMPLATE, HeuristicType.USER_AUTOTUNE)
```
Comment (peterbell10) [inline torch/_inductor/triton_heuristics.py:345]: @oulgen @eellison this leaves fma generation enabled for template kernel and user provided kernels.

**Comment by eellison** [comment]
`eca_halonext26ts` may have just been flakey... rebase ?

**Review by eellison — COMMENTED** [review]
checking in - what is latest state of this pr? were you running into any issues and do we still think this is the best approach

**Comment by eellison** [comment]
sounds good!
